File: 2023/rust/day3/src/main.rs
use pt1::Coord;
use std::{fs, io};

mod pt1;
mod pt2;

enum Part {
    P1,
    P2,
}
const TEST: bool = false;
fn main() -> std::io::Result<()> {
    let dat = read()?;
    run(&dat, Part::P1);
    run(&dat, Part::P2);

    Ok(())
}

fn run(dat: &str, p: Part) {
    let mut nums: Vec<i32> = Vec::new();
    let mut visited: Vec<Coord> = Vec::new();
    // for every line
    for (i, line) in dat.lines().enumerate() {
        // for every char in the line!()
        for (j, c) in line.chars().enumerate() {
            // if the char is a symbol, look around it for numbers
            if !c.is_digit(10) && c != '.' {
                match p {
                    Part::P1 => pt1::run(dat, i, j, &mut nums, &mut visited),
                    Part::P2 => pt2::run(dat, i, j, &mut nums, &mut visited),
                }
            }
        }
    }
    match p {
        Part::P1 => println!(
            "p1: {} ({})",
            nums.iter().sum::<i32>(),
            nums.iter().sum::<i32>() == 536202
        ),
        Part::P2 => println!(
            "p2: {} ({})",
            nums.iter().sum::<i32>(),
            nums.iter().sum::<i32>() == 78272573
        ),
    }
}
fn read() -> io::Result<String> {
    return if !TEST {
        let f = fs::read_to_string("../../data/3.txt")?;
        Ok(f)
    } else {
        let rtn = "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."
            .to_string();
        Ok(rtn)
    };
}

File: 2023/go/day3/main.go
package main

import (
	"fmt"
	"os"
	"strings"
	"unicode"
)

const (
	P1 = iota
	P2
)

func main() {
	data := getData()
	if len(os.Args) > 1 {
		data = `467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..`
	}
	run(data, P1)
	run(data, P2)

}

func run(data string, part uint) {
	lines := strings.Split(data, "\n")
	nums := []int{}
	visited := []Coord{}

	dat := [][]rune{}
	for _, l := range lines {
		dat = append(dat, []rune(l))
		// fmt.Println(l)
	}

	// for every line
	for i, l := range lines {
		//for every rune in the line
		for j, r := range []rune(l) {
			// if the char is a symbol, look around it for numbers
			if !unicode.IsDigit(r) && r != '.' {
				switch part {
				case P1:
					nums, visited = p1(dat, i, j, nums, visited)
				case P2:
					if r == '*' {
						nums, visited = p2(dat, i, j, nums, visited)
					}
				}
			}
		}
	}

	s := 0
	// fmt.Println(nums)
	// fmt.Println(visited)
	for _, n := range nums {
		s += n
	}
	switch part {
	case P1:
		fmt.Printf("p1: %d\n", s)
	case P2:
		fmt.Printf("p2: %d\n", s)
	}

}

func getData() string {
	b, err := os.ReadFile("../../data/3.txt")
	if err != nil {
		panic(err)
	}
	data := string(b)

	return data
}
